### PR TITLE
Custom TF Embedding Layer added

### PR DESCRIPTION
**Pull request by tejas4888**
- Created the first custom trainable TensorFlow layer
- TF Embedding layer (custom init-call) added
- Tests for TF Embedding layer added

### CODE REVIEW

File: box_embeddings/modules/tf_embedding.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            box_initializer = TFUniformBoxInitializer(
+                dimensions=embedding_dim,  # here dim is box dim
+                num_boxes=int(self.input_dim),  # type: ignore
+                box_type_factory=self.box_factory,
+            )
+        self.built = True
+        box_initializer(self.embeddings)
+
+    def call(self, inputs: tf.Tensor) -> TFBoxTensor:
+        emb = super().call(inputs)
```
Comment (dhruvdcoder) [inline box_embeddings/modules/tf_embedding.py:54]: Don't use call explicitly, use the class as a callable. Like: `emb = super()(inputs)`

File: tests/modules/test_tf_embedding.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+def test_forward() -> None:
+    embedding = TFBoxEmbedding(5, 10)
+    box_factory = TFBoxFactory('mindelta_from_vector')
+    W = embedding.embeddings
+    assert W.shape == (5, 20)
+    box_tensor = box_factory(W)
+    box_tensor_direct = box_tensor.data.numpy()[2:4, ...]
+    inputs = tf.Variable([2, 3])
+    box_tensor_from_emb = embedding.call(inputs)
```
Comment (dhruvdcoder) [inline tests/modules/test_tf_embedding.py:33]: Don't use `call` explicitly. Use the class instance as a callable `box_tensor_from_emb = embedding(inputs)`

**Review by dhruvdcoder — CHANGES_REQUESTED** [review]
The embedding layer looks good to me. Just one minor change.

File: box_embeddings/modules/tf_embedding.py (hunk 1 above)
Comment (tejas4888) [reply box_embeddings/modules/tf_embedding.py:54]: I think this call is to the tf.keras.layers.Embedding's call function. To call that we would need to do a super().call(inputs) explicitly else we wont be able to inherit

File: tests/modules/test_tf_embedding.py (hunk 2 above)
Comment (tejas4888) [reply tests/modules/test_tf_embedding.py:33]: Noted. Will make this change. The function signature `call()` in embedding will be replaced by `__call__()` with same functionality

File: box_embeddings/initializations/uniform_boxes.py [hunk 3]
```diff
@@ -111,8 +111,9 @@ def __call__(self, t: torch.Tensor) -> None:  # type:ignore
         )
         with torch.no_grad():
             W = self.box_type_factory.box_subclass.W(z, Z, **self.box_type_factory.kwargs_dict)  # type: ignore
-
+            # print(W.shape,t.shape)
```
Comment (ssdasgupta) [inline box_embeddings/initializations/uniform_boxes.py:114]: should these things be deleted?

Comment (ssdasgupta) [reply box_embeddings/initializations/uniform_boxes.py:114]: I mean the print command for debugging purpose

File: box_embeddings/modules/intersection/tf_gumbel_intersection.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+import numpy as np
+
+
+def _tf_compute_logaddexp_with_clipping_and_separate_forward(
+    t1: TFTBoxTensor, t2: TFTBoxTensor, gumbel_beta: float
+) -> Tuple[tf.Tensor, tf.Tensor]:
+    lse_z = np.logaddexp(t1.z / gumbel_beta, t2.z / gumbel_beta)
+    z = gumbel_beta * (lse_z)
+    z_value = tf.math.maximum(z, tf.math.maximum(t1.z, t2.z))  # type: ignore
+    z_final = (z - tf.stop_gradient(z)) + tf.stop_gradient(z_value)
```
Comment (ssdasgupta) [inline box_embeddings/modules/intersection/tf_gumbel_intersection.py:16]: Clarification question:
why are we using tf.stop_gradient(z) here?

File: box_embeddings/modules/tf_embedding.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        self.num_embeddings = num_embeddings
+        self.embedding_dim = self.output_dim
+        self.box_factory = box_factory
+
+        self.embeddings = self.add_weight(
+            shape=(self.input_dim, self.output_dim),
+            initializer=self.embeddings_initializer,
+            dtype=tf.float64,
+            name='embeddings',
+            regularizer=self.embeddings_regularizer,
```
Comment (ssdasgupta) [inline box_embeddings/modules/tf_embedding.py:39]: what is `self.embeddings_regularizer`

File: box_embeddings/common/tf_utils.py [hunk 6]
```diff
@@ -124,11 +124,11 @@ def softplus_inverse(
     t: tf.Tensor, beta: float = 1.0, threshold: float = 20
 ) -> tf.Tensor:
     below_thresh = beta * t < threshold
-    res = t
+    res_n = t.numpy()
```
[the remark this answers is not in the document]
Comment (tejas4888) [reply box_embeddings/common/tf_utils.py]: Have changed this to `tf.where`

**Comment by tejas4888** [comment]
- Passing `box_initializer` to `self.add_weight` was throwing error. Have made it as a separate call for now
- Volume and Intersection modules are ready for the first release

**Pull request merged**